* DSC530
* Week 3&4 Assignment
* Jeremy Barton

# Hands-on data analysis with Pandas

# Chapter 1, Exercise 4

Run the code in the first cell of the exercises.ipynb notebook. It will give you a list of 100 values to work with for the rest of the exercises in this chapter. Be sure to treat these values as a sample of the population.


In [3]:
import random

random.seed(0)
salaries = [round(random.random()*1000000, -3) for _ in range(100)]
# Currently, this code generated a list of random values as salaries for this exercise.

## Chapter 1, Exercise 5

Using the data from exercise 4, calculate the following statistics without importing anything from the [statistics module](https://docs.python.org/3/library/statistics.html) in the Python standard library, and then confirm your results match up to those that are obtained when using the statistics module (where possible):


Tip from instructor: You can use [Pandas](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html) or [Numpy](https://numpy.org/doc/stable/reference/routines.statistics.html) for the initial calculation and then check against other Python library/package such as [statistics module](https://docs.python.org/3/library/statistics.html), or [SciPy Statistics](https://docs.scipy.org/doc/scipy/reference/stats.html). Part a of Exercise 5 is completed to illustrate the process for you, you can decide how to go about the other problems.

Exercise 5(a) Mean - with exmaples

In [4]:
# Run this code to make alias to shorted the typed name
# You can do this as the 1st thing in any notebook or later, but you only need to do this once
import numpy as np
import pandas as pd

In [5]:
# Using numpy.mean
np.mean(salaries)


np.float64(585690.0)

In [6]:
# Using statistics module
import statistics # no need to re-run this after import
# You have to import this standard library before using its mean function
statistics.mean(salaries)

585690.0

In [7]:
# Using Pandas
# If you choose to use Pandas functions, you have to convert the list to a Series, otherwise it will gives an error
salaries_series = pd.Series(salaries)
# For Pandas function, when you placed a dot after a variable or other functions, it is referred as a "method chaining" appraoch.
salaries_series.mean()

np.float64(585690.0)

In [8]:
# Using SciPy Statistics
from scipy import stats
stats.tmean(salaries)

np.float64(585690.0)

Exercise 5(b) Median

In [9]:
# Your 1st appraoch
print(np.median(salaries))

589000.0


In [10]:
# Using statistics module
print(statistics.median(salaries))

589000.0


Exercise 5(c) Mode

Hint from textbook: Check out the [Counter class](https://docs.python.org/3/library/collections.html#collections.Counter) in the collections module of the standard library

In [11]:
# Your 1st approach
# Returns a list of the n most common elements & their counts
from collections import Counter
Counter(salaries).most_common(1)

[(477000.0, 3)]

In [12]:
# Using statistics module
statistics.mode(salaries)

477000.0

Exercise 5(d) Sample Variance

In [13]:
# Your 1st approach
print(np.var(salaries))

69957413900.0


In [14]:
# Using statistics module
print(statistics.pvariance(salaries))


69957413900.0


Exercise 5(e) Sample Standard Deviation

In [15]:
# Your 1st approach
print(np.std(salaries, ddof=1))

265827.11382484


In [16]:
# Using statistics module
print(statistics.stdev(salaries))

265827.11382484


## Chapter 1, Exercise 6

Using the data from exercise 4, calculate the following statistics using the functions in the statistics module where appropriate:

Exercise 6(a) Range

In [17]:
# Hint:there is no function called "Range". However, what does it means to see a range of values?
# Your solution
max_sal = np.max(salaries)
min_sal = np.min(salaries)
range = max_sal-min_sal
print(range)

995000.0


Exercise 6(b) Coefficient of Variation

In [18]:
# Hint: Coefficient of Variation (CV) is defined as the ratio of standard deviation to the mean in Hands-on Data Analysis with Pandas.
# Your solution
np.std(salaries, ddof=1) / np.mean(salaries)

np.float64(0.45386998894439035)

Exercise 6(c) Interquartile Range

In [19]:
# Hint: Interquartile range (IQR) is presented as the difference of the 75% and 25% quantile in Chapter 1 of Practical Statistics for Data Scientisits Textbook
# OR you can find the values in each quartile and then substract Quantile 3 with Quantile 1 as in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution

q1 = np.quantile(salaries, 0.25)
q3 = np.quantile(salaries, 0.75)

# print(q1)
# print(q3)
IQR = q3-q1
print(f"Interquartile Range:", IQR)

Interquartile Range: 413250.0


Exercise 6(d) Quartile coefficient of dispersion

In [20]:
# Once you determined the 1st and 3rd quantile values from 6(c), you can use it for this calculation
# Refer to the Quartile coefficient of dispersion formula in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution
QCD = (IQR/2)/(q1+q3/2)
print(f"Quartile coefficient of dispersion:", QCD)

Quartile coefficient of dispersion: 0.2545034642032333


## Chapter 1, Exercise 7

Scale the data created in exercise 4 using the following strategies:

Exercise 7(a) Min-max Scaling (Normalizing)

In [21]:
import numpy as np
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the scikit-learn function "MinMaxScaler", which requires transforming the data in 2D array before applying the scaler. This is a common approach in machine learning.
# Your solution
from sklearn.preprocessing import MinMaxScaler

salaries = np.array(salaries)
salaries_scaled = (salaries - np.min(salaries)) / (np.max(salaries) - np.min(salaries))

MMS = MinMaxScaler(salaries)
#print(salaries_scaled)
#print(MMS)

Exercise 7(b) Standardizing

In [22]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the z-score method in Scipy Stats (2D array is needed)
# Your solution

from sklearn.preprocessing import StandardScaler

# Reshape to 2D
salaries = salaries.reshape(-1,1)

scaler = StandardScaler()

standardized_salaries = scaler.fit_transform(salaries)

print("Standardized salaries:", standardized_salaries[:10])

Standardized salaries: [[ 0.97661715]
 [ 0.65146878]
 [-0.62265912]
 [-1.23514791]
 [-0.28238758]
 [-0.68315184]
 [ 0.74976945]
 [-1.06879293]
 [-0.41093461]
 [-0.01017034]]


## Chapter 1, Exercise 8

Using the scaled data from exercise 7, calculate the following:

Note. There is some preparation work to create the normalized (scaled) vs. standardized dataset before calculating covariance.

In [23]:
# If you did not setup a scaled data or standardized data with z-score approach, you can use the following codes.
# Scaled data aka normalized data
from sklearn.preprocessing import MinMaxScaler
# Make an array of salaries
salaries_array = np.array(salaries)
# Reshape the data to 2D (one feature)
salaries_2d = salaries_array.reshape(-1, 1)
# Scaled the data
scaler = MinMaxScaler()
scaler.fit(salaries_2d)
# Optionally, transform the data. scaled_salaries in the scaled data for use in next step
scaled_salaries = scaler.transform(salaries_2d)
# Preview the first 10 values. Noticed the double [] that shows the 2D shape
scaled_salaries[:10]
# This will print the shape if you would like to check
# scaled_salaries.shape

array([[0.84723618],
       [0.76080402],
       [0.42211055],
       [0.25929648],
       [0.51256281],
       [0.40603015],
       [0.78693467],
       [0.30351759],
       [0.47839196],
       [0.58492462]])

In [24]:
# Standardized version of the salary data (z-score)
from scipy import stats
# Assign a new variable `standardized_salary` to store the standarized data
standardized_salary = stats.zscore(salaries_array)
# Preview the first 10 values. This is a 1 dimension object
standardized_salary[:10]
# This will print the shape if you would like to check
# standardized_salary.shape

array([[ 0.97661715],
       [ 0.65146878],
       [-0.62265912],
       [-1.23514791],
       [-0.28238758],
       [-0.68315184],
       [ 0.74976945],
       [-1.06879293],
       [-0.41093461],
       [-0.01017034]])

Exercise 8(a) The covariance between the standardized and normalized data

In [25]:
# Covariance can be calculated with numpy
# When the (normalized) scaled_salaries varaible was created, it was created as a 2D array.
# We will have to 'flatten' it to match the shape of the standardized_salary before calculation.
scaled_salaries = np.array(scaled_salaries).flatten()
standardized_salary = np.array(standardized_salary).flatten()

# Now calculate the covariance
cov_calculated = np.cov(scaled_salaries, standardized_salary)
print(cov_calculated)

[[0.07137603 0.26850885]
 [0.26850885 1.01010101]]


Note: How to interpret this output? The covariance function produce a 2X2 covariance matrix, the off-diagonal elements present the covariance of the 2 sets of values. Therefore 0.2685 is the output for intrepretation. Since the covariance estimate is positive, it indicates a positive relationship (moving in the same direction), and we don't know how "strong" this relationship and therefore correlation will help us understand this better.

In [26]:
# Example 2 for covariance calculation
# Covariance can be calculated with statistics module
statistics.covariance(scaled_salaries, standardized_salary)

np.float64(0.2685088459449036)

Exercise 8(b) The Pearson correlation coefficient between the standardized and normalized data (this is actually 1, but due to rounding along the way, the result will be slightly less)

In [27]:
# Correlation can be calculated with statistics module
statistics.correlation(scaled_salaries, standardized_salary)
# The result is 1 because we are checking relationship of the same data in different format.

np.float64(1.0000000000000002)

Note. Getting a correlation of 1 is often very rare, but in this case, we are basically using the same raw data and simply normalized it (force values to 0,1 range) or standardized it (force the data to centered around zero and then deviate from zero) to put it on a common scale. It will not change the strength or direction of the relationship between 2 variables.

# Chapter 2

In [28]:
# The default code to run if you setup your book_env correctly to import the dataset
earthquake_data = pd.read_csv('ch_02/data/parsed.csv')

# If you encountered issues due to directory setup, you can provide the full file path also
# earthquake_data = pd.read_csv('cdrive../../ch_02/data/parsed.csv')

## Chapter 2, Exercise 1

Find the 95th percentile of earthquake magnitude in Japan using the mb magnitude type.

In [ ]:
import pandas as pd
import numpy as np

# Filtering Japan with magType 'mb'
japan_mb = earthquake_data[
    (earthquake_data['place'].str.contains('Japan', case=False, na=False)) & 
    (earthquake_data['magType'] == 'mb') & 
    (earthquake_data['mag'].notna())
]

# Compute the 95th percentile
percentile_95 = np.percentile(japan_mb['mag'], 95)
print(percentile_95)


4.9


## Chapter 2, Exercise 2

Find the percentage of earthquakes in Indonesia that were coupled with tsunamis.

In [30]:
import pandas as pd

# Reload dataset
earthquake_data = pd.read_csv('ch_02/data/earthquakes.csv')

indo_quakes = earthquake_data[earthquake_data['place'].str.contains('Indonesia', case=False, na=False)]

# Tsunami quake count
tsunami_quakes = indo_quakes[indo_quakes['tsunami'] == 1]

# Calculate percentage
p = (len(tsunami_quakes) / len(indo_quakes)) * 100

print(f"Percentage of earthquakes in Indonesia with tsunamis: {p:.2f}%")


Percentage of earthquakes in Indonesia with tsunamis: 23.13%


## Chapter 2, Exercise 3

Calculate summary statistics for earthquakes in Nevada.

In [31]:
nv_quakes = earthquake_data[earthquake_data['place'].str.contains('Nevada', case=False, na=False)]
nv_quakes.describe

<bound method NDFrame.describe of      alert  cdi      code                                             detail  \
32     NaN  NaN  00660878  https://earthquake.usgs.gov/fdsnws/event/1/que...   
39     NaN  NaN  00660886  https://earthquake.usgs.gov/fdsnws/event/1/que...   
41     NaN  NaN  00660885  https://earthquake.usgs.gov/fdsnws/event/1/que...   
48     NaN  NaN  00660867  https://earthquake.usgs.gov/fdsnws/event/1/que...   
49     NaN  NaN  00660884  https://earthquake.usgs.gov/fdsnws/event/1/que...   
...    ...  ...       ...                                                ...   
9191   NaN  NaN  00657063  https://earthquake.usgs.gov/fdsnws/event/1/que...   
9199   NaN  NaN  00657203  https://earthquake.usgs.gov/fdsnws/event/1/que...   
9220   NaN  NaN  00657032  https://earthquake.usgs.gov/fdsnws/event/1/que...   
9245   NaN  NaN  00657199  https://earthquake.usgs.gov/fdsnws/event/1/que...   
9253   NaN  NaN  00657031  https://earthquake.usgs.gov/fdsnws/event/1/que...   

     

## Chapter 2, Exercise 4

Add a column indicating whether the earthquake happened in a country or US state that is on the Ring of Fire. Use Alaska, Antarctica (look for Antarctic), Bolivia, California, Canada, Chile, Costa Rica, Ecuador, Fiji, Guatemala, Indonesia, Japan, Kermadec Islands, Mexico (be careful not to select New Mexico), New Zealand, Peru, Philippines, Russia, Taiwan, Tonga, and Washington.

In [32]:
RoF = r'\b(Alaska|Antarctica|Antarctic|Bolivia|California|Canada|Chile|Costa Rica|' \
          r'Ecuador|Fiji|Guatemala|Indonesia|Japan|Kermadec Islands|Mexico(?!\s*New)|' \
          r'New Zealand|Peru|Philippines|Russia|Taiwan|Tonga|Washington)\b'


# A negative lookbehind helps to avoid picking up New Mecixo

# Create column with True/False if place matches Ring of Fire location
earthquake_data['on_ring_of_fire'] = earthquake_data['place'].str.contains(RoF, case=False, na=False, regex=True)

# Check
# earthquake_data[['place', 'on_ring_of_fire']]

/tmp/ipykernel_15208/4167205334.py:9: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  earthquake_data['on_ring_of_fire'] = earthquake_data['place'].str.contains(RoF, case=False, na=False, regex=True)


## Chapter 2, Exercise 5

Calculate the number of earthquakes in the Ring of Fire locations and the number outside of them.

In [33]:
in_RoF = earthquake_data['on_ring_of_fire'].sum()
not_in_rof =len(earthquake_data) - in_RoF

print(f"Earthquakes in RoF:", in_RoF)
print(f"Outside RoF:", not_in_rof)



Earthquakes in RoF: 4427
Outside RoF: 4905


## Chapter 2, Exercise 6

Find the tsunami count along the Ring of Fire.

In [34]:
tsunami_in_RoF = earthquake_data[
    (earthquake_data['on_ring_of_fire'] == True) & 
    (earthquake_data['tsunami'] == 1)
]

tsunami_count = len(tsunami_in_RoF)
print(f"Tsunamis along RoF:", tsunami_count)

Tsunamis along RoF: 43
